services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java

```diff
@@ -0,0 +1,339 @@
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.protobuf.service;
+
+import com.google.common.base.Preconditions;
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.Descriptors.FieldDescriptor;
+import com.google.protobuf.Descriptors.FileDescriptor;
+import com.google.protobuf.Descriptors.MethodDescriptor;
+import com.google.protobuf.Descriptors.ServiceDescriptor;
+
+import io.grpc.ExperimentalApi;
+import io.grpc.InternalNotifyOnServerBuild;
+import io.grpc.Server;
+import io.grpc.ServerServiceDefinition;
+import io.grpc.Status;
+import io.grpc.protobuf.ProtoFileDescriptorSupplier;
+import io.grpc.reflection.v1alpha.ErrorResponse;
+import io.grpc.reflection.v1alpha.ExtensionNumberResponse;
+import io.grpc.reflection.v1alpha.ExtensionRequest;
+import io.grpc.reflection.v1alpha.FileDescriptorResponse;
+import io.grpc.reflection.v1alpha.ListServiceResponse;
+import io.grpc.reflection.v1alpha.ServerReflectionGrpc;
+import io.grpc.reflection.v1alpha.ServerReflectionRequest;
+import io.grpc.reflection.v1alpha.ServerReflectionResponse;
+import io.grpc.reflection.v1alpha.ServiceResponse;
+import io.grpc.stub.StreamObserver;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ * Provides a reflection service for Protobuf services (including the reflection service itself).
+ *
+ * <p>Throws an exception if the set of services contain multiple protobuf files with declarations
+ * of the same service, method, type, or extension.
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/2222")
+public final class ProtoReflectionService extends ServerReflectionGrpc.ServerReflectionImplBase
+    implements InternalNotifyOnServerBuild {
+
+  private Server server;
+
+  /**
+   * Retrieves the services registered to the server at build time, and stores those services with a
+   * {@link ProtoFileDescriptorSupplier}. Subsequent calls to the method will have no effect.
+   */
+  @Override
+  public void notifyOnBuild(Server server) {
+    if (this.server == null) {
+      this.server = Preconditions.checkNotNull(server, "server");
+    }
+  }
+
+  @Override
+  public StreamObserver<ServerReflectionRequest> serverReflectionInfo(
```

Comment:
Is there an example anywhere that uses both `disableAutoInboundFlowControl()` and `setOnReadyHandler()`? I could not find one. How I wanted to do this was:

```
final ServerCallStreamObserver<ServerReflectionResponse> serverCallObserver =
    (ServerCallStreamObserver<ServerReflectionResponse>) responseObserver;

serverCallObserver.disableAutoInboundFlowControl();
serverCallObserver.request(1);

serverCallObserver.setOnReadyHandler(new Runnable() {
  @Override
  public void run() {
    while (serverCallObserver.isReady()) {
      serverCallObserver.request(1);
    }
  }
});
```

This doesn't work - the onReadyHandler runs in the same thread as the requestObserver, so the while loop just runs forever and nothing happens. What does work is removing the `while` loop and just requesting 1 message every time the stream observer becomes ready. This then requires the `onNext()` method in the `StreamObserver<ServerReflectionRequest>` to also check if `serverCallObserver` is ready and, if so, request another message.

I feel like this is not the intended use of the flow control API. What would be nice is to actually do the work inside of the `onReadyHandler`, e.g.:

```
serverCallObserver.setOnReadyHandler(new Runnable() {
  @Override
  public void run() {
    while (serverCallObserver.isReady()) {
      serverCallObserver.onNext(<some reflection response>);
      serverCallObserver.request(1);
    }
  }
});
```

This is similar to @ejona86's comment from Sept 9th on https://github.com/grpc/grpc-java/issues/2247, except there the message passed to `onNext()` did not depend on anything coming in from the client. Essentially, I haven't been able to figure out how to access the `ReqT request` that we get [here](https://github.com/grpc/grpc-java/blob/master/stub/src/main/java/io/grpc/stub/ServerCalls.java#L206) within the `onReadyHandler`, except by the hack described above.

Commit https://github.com/grpc/grpc-java/commit/7902017b07d09ccff6d118965b9de804d37960b0 introduced the disableAutoInboundFlowControl() and references `TestServiceImpl.streamingOutputCallManualFlowControl` as an example of the usage, but I can't find that method anywhere in Github.